invoices/invoiceregistry_test.go

```diff
@@ -150,6 +150,109 @@ func TestSettleInvoice(t *testing.T) {
 	if inv.AmtPaid != amtPaid {
 		t.Fatal("expected amount to be unchanged")
 	}
+
+	// Try to cancel.
+	err = registry.CancelInvoice(hash)
+	if err == nil {
+		t.Fatal("expected cancelation of a settled invoice to fail")
+	}
+}
+
+// TestCancelInvoice tests cancelation of an invoice and related notifications.
+func TestCancelInvoice(t *testing.T) {
+	cdb, cleanup, err := newDB()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer cleanup()
+
+	// Instantiate and start the invoice registry.
+	registry := NewRegistry(cdb, &chaincfg.MainNetParams)
+
+	err = registry.Start()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer registry.Stop()
+
+	allSubscriptions := registry.SubscribeNotifications(0, 0)
+	defer allSubscriptions.Cancel()
+
+	// Subscribe to the not yet existing invoice.
+	subscription := registry.SubscribeSingleInvoice(hash)
+	defer subscription.Cancel()
+
+	if subscription.hash != hash {
+		t.Fatalf("expected subscription for provided hash")
+	}
+
+	// Add the invoice.
+	invoice := &channeldb.Invoice{
+		Terms: channeldb.ContractTerm{
+			PaymentPreimage: preimage,
+			Value:           lnwire.MilliSatoshi(100000),
+		},
+	}
+
+	_, err = registry.AddInvoice(invoice, hash)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// We expect the open state to be sent to the single invoice subscriber.
+	select {
+	case update := <-subscription.Updates:
+		if update.Terms.State != channeldb.ContractOpen {
+			t.Fatalf(
+				"expected state ContractOpen, but got %v",
+				update.Terms.State,
+			)
+		}
+	case <-time.After(testTimeout):
+		t.Fatal("no update received")
+	}
+
+	// We expect a new invoice notification to be sent out.
+	select {
+	case newInvoice := <-allSubscriptions.NewInvoices:
+		if newInvoice.Terms.State != channeldb.ContractOpen {
+			t.Fatalf(
+				"expected state ContractOpen, but got %v",
+				newInvoice.Terms.State,
+			)
+		}
+	case <-time.After(testTimeout):
+		t.Fatal("no update received")
+	}
+
+	// Cancel invoice.
+	err = registry.CancelInvoice(hash)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// We expect the canceled state to be sent to the single invoice
+	// subscriber.
+	select {
+	case update := <-subscription.Updates:
+		if update.Terms.State != channeldb.ContractCanceled {
+			t.Fatalf(
+				"expected state ContractCanceled, but got %v",
+				update.Terms.State,
+			)
+		}
+	case <-time.After(testTimeout):
+		t.Fatal("no update received")
+	}
+
+	// We expect no cancel notification to be sent to all invoice
+	// subscribers (backwards compatibility).
+
+	// Try to cancel again.
+	err = registry.CancelInvoice(hash)
+	if err != nil {
+		t.Fatal("expected cancelation of a canceled invoice to succeed")
+	}
```

Comment:
two other cases that we might consider adding:
 - canceling an unknown payment hash
 - settling an invoice that's already been canceled

Reply:
Unknown payment hash is something that can never happen. Every invoice has a payment hash. Unknown preimage can happen, but not yet in this PR.

Reply:
Settling a canceled invoice added to the test.

Reply:
> Unknown payment hash is something that can never happen

How so? What happens if i call `CancelInvoice`on this invoice registry with an unknown payment hash? I should get [this error](https://github.com/lightningnetwork/lnd/pull/2457/files#diff-9f84ce21f17bb10010b2d2b19cb17e3cR673) returned [here](https://github.com/lightningnetwork/lnd/pull/2457/files#diff-4c01bfee90d73adb114602835d044d8dR491), no?

Reply:
Oh yes, ofc. Misunderstood you.

Reply:
Test added.